BUG: `dtype` not being preserved for `replace` on a `CategoricalDtype`

### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.


### Reproducible Example

```python
>>> import pandas as pd
>>> pd.__version__
'1.4.2'
>>> pdf = pd.DataFrame(
...             {
...                 "a": ["one", "two", None, "three"],
...                 "b": ["one", None, "two", "three"],
...             },
...             dtype="category",
...         )
>>> pdf.dtypes
a    category
b    category
dtype: object
>>> pdf
       a      b
0    one    one
1    two    NaN
2    NaN    two
3  three  three
>>> new_df = pdf.replace(to_replace=[".", "def"], value=["_", None])
>>> new_df
       a      b
0    one    one
1    two    NaN
2    NaN    two
3  three  three
>>> new_df.dtypes
a    object
b    object
dtype: object
```


### Issue Description

When the series is of `category` dtype, calling `replace` with above inputs is not preserving the dtype.

### Expected Behavior

```python
>>> import pandas as pd
>>> pd.__version__
'1.3.5'
>>> pdf = pd.DataFrame(
...             {
...                 "a": ["one", "two", None, "three"],
...                 "b": ["one", None, "two", "three"],
...             },
...             dtype="category",
...         )
>>> new_df = pdf.replace(to_replace=[".", "def"], value=["_", None])
>>> new_df
       a      b
0    one    one
1    two    NaN
2    NaN    two
3  three  three
>>> new_df.dtypes
a    category
b    category
dtype: object
```

### Installed Versions

<details>

>>> pd.show_versions()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/pandas/util/_print_versions.py", line 109, in show_versions
    deps = _get_dependency_info()
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/pandas/util/_print_versions.py", line 88, in _get_dependency_info
    mod = import_optional_dependency(modname, errors="ignore")
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/pandas/compat/_optional.py", line 138, in import_optional_dependency
    module = importlib.import_module(name)
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/setuptools/__init__.py", line 8, in <module>
    import _distutils_hack.override  # noqa: F401
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/_distutils_hack/override.py", line 1, in <module>
    __import__('_distutils_hack').do_override()
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/_distutils_hack/__init__.py", line 72, in do_override
    ensure_local_distutils()
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/_distutils_hack/__init__.py", line 59, in ensure_local_distutils
    assert '_distutils' in core.__file__, core.__file__
AssertionError: /nvme/0/pgali/envs/cudfdev/lib/python3.8/distutils/core.py

</details>

--- a/pandas/tests/frame/methods/test_replace.py
+++ b/pandas/tests/frame/methods/test_replace.py
@@ -1559,3 +1559,17 @@ class TestDataFrameReplaceRegex:
         result = df.replace({0: 1, 1: np.nan})
         expected = DataFrame({"A": [1, np.nan, 2], "B": [np.nan, 1, 2]})
         tm.assert_frame_equal(result, expected)
+
+    def test_replace_categorical_no_replacement(self):
+        # GH#46672
+        df = DataFrame(
+            {
+                "a": ["one", "two", None, "three"],
+                "b": ["one", None, "two", "three"],
+            },
+            dtype="category",
+        )
+        expected = df.copy()
+
+        result = df.replace(to_replace=[".", "def"], value=["_", None])
+        tm.assert_frame_equal(result, expected)

